src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java

```diff
@@ -0,0 +1,224 @@
+package fi.aalto.cs.apluscourses.intellij.psi;
+
+import com.intellij.psi.PsiElement;
+import com.intellij.psi.PsiParameter;
+import com.intellij.psi.PsiWhiteSpace;
+import com.intellij.psi.impl.source.tree.LeafPsiElement;
+import fi.aalto.cs.apluscourses.utils.ArrayUtil;
+import fi.aalto.cs.apluscourses.utils.StringUtil;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.jetbrains.plugins.scala.lang.psi.api.base.ScAnnotations;
+import org.jetbrains.plugins.scala.lang.psi.api.base.ScPrimaryConstructor;
+import org.jetbrains.plugins.scala.lang.psi.api.base.types.ScSimpleTypeElement;
+import org.jetbrains.plugins.scala.lang.psi.api.statements.params.ScTypeParamClause;
+import org.jetbrains.plugins.scala.lang.psi.impl.base.ScModifierListImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScTemplateParentsImpl;
+
+public class ScalaClassDeclaration {
```

Comment:
If we can create PSI elements from string literals (somehow like `PsiElement elem = PsiElement.fromString("class Foo() exteds Bar {}")`), this class could be easily and thoroughly unit-tested.

Reply:
There is [this](https://github.com/JetBrains/intellij-community/blob/master/java/java-psi-impl/src/com/intellij/psi/impl/PsiElementFactoryImpl.java) promising class.